jet_tracking/datastream.py

```diff
@@ -122,370 +163,414 @@ def read_value(self):  # needs to initialize first maybe using a decorator?
-                old = list(b[key])
-                new = old[-d:]
-                b[key] = deque(new, new_size)
+        This method stops the StatusThread by setting the paused attribute to True. If abort is True, the thread will be interrupted.
+
+        Parameters:
+        ----------
+        abort : bool
+            A flag indicating whether the thread should be interrupted.
+        """
+        self.paused = True
+        if abort:
+            self.thread().requestInterruption()
         else:
-            for key in b:
-                old = list(b[key])
-                b[key] = deque(old, new_size)
-        return b
+            pass
 
-    @staticmethod
-    def append_or_set(choice, d, vals, idx):
+    def run_data_thread(self):
         """
-        Appends or assigns values to a dictionary.
+        Run the data thread.
 
-        Send in a list of vals in the order 'i0', 'diff', 'ratio' and 'time'
-        and the choice to either "append" or "set" and the dictionary d and
-        then execute the action.
+        This method is the long-running task that collects data points. It reads values from the ValueReader,
+        updates buffers, checks status updates, calculates averages, and emits signals to refresh graphs.
 
-        Keyword arguments:
-        choice -- either "append" or "set"
-        d -- the dictionary that will be updated
-        vals -- the values or value to be added
-        idx -- the current x_cycle value
+        Note: This method runs in a continuous loop until the thread is paused.
         """
-
-        keys = list(d.keys())
-        if choice == "append":
-            for key, val in list(zip(keys, vals)):
-                d[key].append(val)
-        elif choice == "set":
-            for key, val in list(zip(keys, vals)):
-                d[key][idx] = val
+        if not self.paused:
+            self.current_values = self.reader.read_value()
+            vals = self.current_values
+            if self.mode == "running":
+                self.update_buffer(vals)
+                self.check_status_update()
+                if self._ave_count == self.averaging_size:
+                    self.calculate_averages()
+                    vals['ratio'] = [vals['ratio'], self.averages['ratio'][-1]]
+                else:
+                    vals['ratio'] = [vals['ratio'], np.nan]
+                self.signals.refreshGraphs.emit(vals, self._count)
+            elif self.mode == "calibrate":
+                self.calibrated = False
+                self.update_buffer(vals)
+                self.check_status_update()
+                self.calibrate(vals)
+                if self._ave_count == self.averaging_size:
+                    self.calculate_averages()
+                    vals['ratio'] = [vals['ratio'], self.averages['ratio'][-1]]
+                else:
+                    vals['ratio'] = [vals['ratio'], np.nan]
+                self.signals.refreshGraphs.emit(vals, self._count)
```

Comment:
There's some confusing mixing of responsibilities in this module.
The data stream here has a huge burden of knowledge: rather than emitting signals related to what this object itself is doing ("here is my data") it instead has to know what its data is being used for ("update the graphs using my data"). As a consequence, developers digging into this module also need to know huge amounts of context to parse and tweak the behavior.

Reply:
This is related to the problems that come along with passing large amounts of global signals.